### ISSUE TITLE
Requiring ILogger<> during function startup throws InvalidOperationException

### ISSUE THREAD

**Issue by jcageman** [Azure/azure-functions-host#5912]
I would like to initialize some caching code within an azure function. This caching service injects an ILogger<> which cannot be injected apparently, since not all required services have been registered.

When running the following line of code (part of the code below). i get the following exception:
`scope.ServiceProvider.GetRequiredService<ICachingService>().Initialize();`

```
An exception of type 'System.InvalidOperationException' occurred in Microsoft.Extensions.DependencyInjection.dll but was not handled in user code
Unable to resolve service for type 'Microsoft.Azure.WebJobs.Script.IFileLoggingStatusManager' while attempting to activate 'Microsoft.Azure.WebJobs.Script.Diagnostics.HostFileLoggerProvider'.
```
#### Investigative information
- Function App version: v3 (Microsoft.NET.Sdk.Functions" Version="3.0.6")
- dot net core 3.1

Startup code:
```
    public class Startup : FunctionsStartup
    {
        public override void Configure(IFunctionsHostBuilder builder)
        {
            builder.Services.AddLogging();
            RegisterServices(builder.Services);
            var serviceProvider = builder.Services.BuildServiceProvider(true);
            InitializeServices(serviceProvider);
        }

        private static void RegisterServices(IServiceCollection services)
        {
            var serviceProvider = services.BuildServiceProvider();
            services.AddSingleton<ICachingService, CachingService>();
        }

        private static void InitializeServices(IServiceProvider serviceProvider)
        {
            using var scope = serviceProvider.CreateScope();
            scope.ServiceProvider.GetRequiredService<ICachingService>().Initialize().Wait();
        }
    }
```

Caching service:
```
    public interface ICachingService
    {
        Task Initialize();
    }

    public class CachingService : ICachingService
    {
        private readonly ILogger<CachingService> _logger;

        public CachingService(ILogger<CachingService> logger)
        {
            _logger = logger;
        }
        public async Task Initialize()
        {
            _logger.LogInformation("initialization");
            await Task.CompletedTask;
        }
    }
```

**Comment by Camios** [comment]
Yes, I have a similar problem where I need a logger (doesn't need to be scoped) for an explicitly constructed service.

**Comment by soninaren** [comment]
@fabiocav , @brettsam any ideas here?

**Comment by brettsam** [comment]
The code above is attempting to resolve a dependency before all dependencies are registered. I'd suggest trying to do the initialization in a factory instead, which will delay construction until all services are registered and the host is ready to start.

Example:
```csharp
public override void Configure(IFunctionsHostBuilder builder)
{
    builder.Services.AddSingleton<ICachingService>(p =>
    {
        var logger = p.GetService<ILogger<CachingService>>();
        var cachingService = new CachingService(logger);
        cachingService.Initialize().Wait();
        return cachingService;
    });
}
```

**Comment by Camios** [comment]
What dependencies aren't resolved for logging by the time the OP attempts to resolve it? Seemes backwards for logging dependencies to registered after the FunctionStartup.Configure methods is called.

**Comment by v-anvari** [comment]
Hi @jcageman , Closing this issue as the workaround was provided and no recent activity. Please feel free to reach back to us in case of any issue

**Comment by Camios** [comment]
The activity we are waiting on is for the Functions team to fix the flaw. The lack of activity is from the MS team. Please reopen and fix it.

**Comment by v-anvari** [comment]
Hi @Camios , Thank you for the response. As this is an old issue, request you to kindly open a new issue with details of your project scenario, so that we can add it to the sprint for further investigation

**Comment by jcageman** [comment]
It's pretty annoying, but I've created a new issue. See #7322
